src/cast_core/src/SamplingLoader.cpp

```diff
@@ -0,0 +1,151 @@
+
+    mongocxx::database database;
+    mongocxx::collection collection;
+    int64_t insertBatchSize;
+    int64_t numBatches;
+    int64_t sampleSize;
+    // User can specify any pipeline here, but it's worth noting that some will not play nicely. For
+    // example, adding a $match will deflate the sample size.
+    PipelineGenerator pipelineSuffixGenerator;
+};
+
+void genny::actor::SamplingLoader::run() {
+    for (auto&& config : _loop) {
+        for (auto&& _ : config) {
+            BOOST_LOG_TRIVIAL(debug) << "Beginning to run SamplingLoader";
+            // Read the sample size documents into a vector.
+            mongocxx::pipeline samplePipeline;
+            samplePipeline.sample(config->sampleSize);
+            samplePipeline.project(make_document(kvp("_id", 0)));
+            const auto suffixStages =
+                pipeline_helpers::makePipeline(config->pipelineSuffixGenerator).view_array();
+            samplePipeline.append_stages(suffixStages);
+            auto cursor =
+                config->collection.aggregate(samplePipeline, mongocxx::options::aggregate{});
+            const auto sampleDocs =
+                std::vector<bsoncxx::document::value>(cursor.begin(), cursor.end());
+
+            if (sampleDocs.empty()) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                    "Sample was unable to find any documents from collection '" +
+                    to_string(config->collection.name()) +
+                    "'. Could the collection be empty or could the pipeline be filtering out "
+                    "documents? Attempting to sample " +
+                    boost::to_string(config->sampleSize) +
+                    " documents. Pipeline suffix = " + to_json(suffixStages)));
+            }
+            if (sampleDocs.size() < config->sampleSize) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                    "Could not get a sample of the expected size. Either the collection '" +
+                    to_string(config->collection.name()) +
+                    "' is smaller than the requested sample size of " +
+                    boost::to_string(config->sampleSize) +
+                    " documents, or the specified pipeline suffix is filtering documents. Found "
+                    "only " +
+                    boost::to_string(sampleDocs.size()) +
+                    " documents. Pipeline suffix = " + to_json(suffixStages)));
+            }
+
+            // Maintain an index into the sample that's used for the insert batches so we
+            // insert roughly the same number of copies for each sample.
+            size_t sampleIdx = 0;
+
+            auto batchOfDocs = std::vector<bsoncxx::document::view_or_value>{};
+            auto totalOpCtx = _totalBulkLoad.start();
+            for (size_t batch = 0; batch < config->numBatches; ++batch) {
+                batchOfDocs.clear();
```

Comment:
reserve() the expected size of the batch? (I don't know whether C++ compilers shrink the capacity of the vector on clear. Another option would be not to clear but overwrite each element in the vector with a new doc because the size of the batch is const.)

Reply:
Good idea! I've adjusted to re-use the memory.